### PR TITLE
hudson plot

### PR DESCRIPTION
**Pull request by braunfuss**

### CODE REVIEW

**Comment by braunfuss** [comment]
Please check if any more plot options should be handed down. Also we might want to change the random number of models drawn to a fixed number from the last models in the chain?

File: src/plotting.py [hunk 1]
```diff
@@ -1485,6 +1485,165 @@ def draw_fuzzy_beachball(problem, po):
...
+    beachballsize = markersize
+    beachballsize_small = beachballsize * 0.5
+
+    fig = plt.figure(figsize=(4., 4.))
+    fig.subplots_adjust(left=0., right=1., bottom=0., top=1.)
+    axes = fig.add_subplot(1, 1, 1)
+
+    data = []
+    for x in m6s:
+        mt = mtm.MomentTensor(
```
Comment (hvasbath) [inline src/plotting.py]: please use instead:
mt = mtm.as_mt(m6)

File: src/plotting.py [hunk 2]
```diff
@@ -1485,6 +1485,165 @@ def draw_fuzzy_beachball(problem, po):
...
+                beachball_type=beachball_type,
+                position=(u, v),
+                size=beachballsize,
+                color_t=color,
+                zorder=2,
+                linewidth=0.5)
+        except beachball.BeachballError as e:
+            logger.warn(str(e))
+
+        mt = mtm.MomentTensor(
```
Comment (hvasbath) [inline src/plotting.py]: see above

File: src/plotting.py [hunk 3]
```diff
@@ -1485,6 +1485,165 @@ def draw_fuzzy_beachball(problem, po):
...
+
+    else:
+        llk_str = 'ref'
+        m6s = num.empty((1, 6), dtype='float64')
+        for i, varname in enumerate(
+                ['mnn', 'mee', 'mdd', 'mne', 'mnd', 'med']):
+            m6s[:, i] = po.reference[varname].ravel()
+
+    m6s_mean = num.mean(m6s, axis=0)
+    m6s_best = m6s[-1, :]
```
Comment (hvasbath) [inline src/plotting.py]: the solutions are not sorted, this wont be the best, see get_result_point how to find the best

File: src/plotting.py [hunk 4]
```diff
@@ -1485,6 +1485,165 @@ def draw_fuzzy_beachball(problem, po):
...
+                varname, combine=True, squeeze=True).ravel()
+
+    else:
+        llk_str = 'ref'
+        m6s = num.empty((1, 6), dtype='float64')
+        for i, varname in enumerate(
+                ['mnn', 'mee', 'mdd', 'mne', 'mnd', 'med']):
+            m6s[:, i] = po.reference[varname].ravel()
+
+    m6s_mean = num.mean(m6s, axis=0)
```
Comment (hvasbath) [inline src/plotting.py]: see get_result_point how to get the mean, we want to have the mean llk, the mean of all the values may aven be an MT we havent sampled yet- we dont want that!

File: src/plotting.py [hunk 5]
```diff
@@ -1485,6 +1485,165 @@ def draw_fuzzy_beachball(problem, po):
...
+                color_t='red',
+                zorder=2,
+                linewidth=0.5)
+        except beachball.BeachballError as e:
+            logger.warn(str(e))
+
+    outpath = os.path.join(
+        problem.outfolder,
+        po.figure_dir,
+        'hudson_%i_%s.%s' % (po.load_stage, llk_str, 'png'))
```
Comment (hvasbath) [inline src/plotting.py]: instead of 'png' please use po.format or doesnt it work with others? then keep png

**Comment by braunfuss** [comment]
Thanks for the comments, very helpful! Should be compliant now.

File: src/plotting.py [hunk 6]
```diff
@@ -1485,6 +1485,163 @@ def draw_fuzzy_beachball(problem, po):
...
+        llk_str = po.post_llk
+        stage = Stage(homepath=problem.outfolder)
+
+        stage.load_results(
+            varnames=problem.varnames,
+            model=problem.model, stage_number=po.load_stage,
+            load='trace', chains=[-1])
+
+        n_mts = len(stage.mtrace)
+        m6s = num.empty((n_mts, 6), dtype='float64')
```
Comment (hvasbath) [inline src/plotting.py]: U use this block 4 times, please make it a function point2array which takes the variable names as an argument to determine the column order of the resulting array

File: src/plotting.py [hunk 7]
```diff
@@ -1485,6 +1485,163 @@ def draw_fuzzy_beachball(problem, po):
...
+                color=color,
+                ms=markersize_small,
+                mec='none',
+                mew=0,
+                alpha=0.25,
+                zorder=0)
+
+        hudson.draw_axes(axes)
+
+        mt = mtm.as_mt(m6_mean)
```
[the remark this answers is not in the document]
Comment (braunfuss) [reply src/plotting.py]: The max and mean beachballs have different colors, also the max one gets a rectangle around, so it is not a repetition.

File: src/plotting.py [hunk 8]
```diff
@@ -1485,6 +1485,163 @@ def draw_fuzzy_beachball(problem, po):
...
+            axes.plot(
+                u, v, 'o',
+                color=color,
+                ms=markersize_small,
+                mec='none',
+                mew=0,
+                alpha=0.25,
+                zorder=0)
+
+        hudson.draw_axes(axes)
```
[the remark this answers is not in the document]
Comment (braunfuss) [reply src/plotting.py]: You are right

File: src/plotting.py (hunk 6 above)
Comment (braunfuss) [reply src/plotting.py]: Like this you mean?
def point2array(a, n_mts=None):

    if n_mts is not None:
        if str(a) is 'm6s':
            return num.empty((n_mts, 6), dtype='float64')
            
    if str(a) is 'm6s':
            return num.empty((1, 6), dtype='float64')

    if str(a) is 'm6_mean' or 'm6_max':
            return num.empty((6), dtype='float64')

**Pull request merged**